### PR TITLE
Introduced SubclassResolver.

### PR DESCRIPTION
**Pull request by jducoeur**
Includes unit tests, and a couple of Serializers designed specifically for use with this Resolver.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -212,6 +212,21 @@ extensions = ["com.romix.akka.serialization.kryo.KryoSerializationExtension$"]
...
+			# problem is that a large number of common Scala and Akka types (such as
+			# Map and ActorRef) are actually traits that mask a large number of
+			# specialized classes that deal with various situations and optimizations.
+			# It isn't straightforward to register all of these, so you can instead
+			# register a single supertype, with a serializer that can handle *all* of
+			# the subclasses, and the subclasses get serialized with that.
+			#
+			# Use this with care: you should only rely on this when you are confident
+			# that the superclass serializer covers all of the special cases properly. 
+			resolve-subclasses = false
```
Comment (romix) [inline README.md]: Please change the indentation of the comments block above to match the rest of the comments before and after it. It seems to be shifted to the right a bit too much.

**Comment by romix** [comment]
Thanks for submitting the PR. Overall, it looks good, except for the minor cosmetic issue I mentioned above.

Please fix it, let me run the build and test, and if everything is OK, I'll merge it.

**Comment by jducoeur** [comment]
*Sigh* -- bloody tabs strike again.  Fixed...

**Comment by romix** [comment]
Thanks! Could you squash the commits and push --force it? It is silly to have a separate commit just for the sake of indentation.

**Comment by jducoeur** [comment]
... in principle, I totally agree, but I'm having difficulty doing this in practice.  I made the mistake of making the indentation change directly in the GitHub UI, rather than in my local repo, and now I can't figure out how to make the squash work.  (Possibly because SmartGit is shielding me from the necessary commands.)

Sorry -- after several years of using git in daily usage, I'm still a bit of a noob when it comes to the ins and out of getting it right for open source projects.  If you need this squashed properly, I may have to tear down the branch and try the PR again with a new one; I'm not sure how to recover the current situation as it is...

**Comment by romix** [comment]
No problem. Let's have 2 commits. It is not the end of the world ;-)

**Comment by romix** [comment]
I tested and all test are green. Merging. Thanks again!

**Comment by jducoeur** [comment]
Cool -- thanks!

**Comment by guidomedina** [comment]
This is an interesting contribution, I only found out few days ago as I have noticed that it would register classes referenced by explicit listed classes?

Can you tell if this would also work for automatic? meaning, is it tied to the ID strategy directly or to the classes registration phase which can be called from explicit or automatic?

**Comment by guidomedina** [comment]
So, basically, does `resolve-subclasses = true` affects also `idstrategy=automatic`?
If that is the case, the documentation should be corrected to say so, if not, I think this functionality should be wider in that it should affect any ID strategy that accepts a list of explicit classes to register which in this case are:
- explicit
- automatic

WDYT?

**Comment by jducoeur** [comment]
As currently coded, I believe it can be turned on with anything other than `incremental`, but I suspect (admittedly, in the morning before I've had my coffee) that it is kind of dangerous with `automatic` -- at least, I have much less confidence in it.

Consider: the point here is that you can explicitly register a supertype, using a serializer that is designed to handle the class hierarchy below it, and it will then automatically deal with those.  Now think about the typical case in `automatic`, where you *aren't* doing consistent explicit registration.  The results are going to be haphazard, depending on the order in which types are received, and might be inconsistent -- it could wind up sometimes working, because you hit a superclass with a registered default serializer first, and sometimes not when you hit a subclass first.

Moreover, it may get pulled automatically into cases it wasn't designed for.  The intended use cases are pretty much all ones where the type you want to register is a *trait* implemented by a bunch of subclasses, all of which can and should get serialized identically.  If you use it in `automatic`, it might get used semi-randomly for superclasses and subclasses, and I'm not sure the results are always going to make sense.  (They might, but it needs analysis.)

Basically, SubclassResolver was written specifically for `explicit` because it is *necessary* for `explicit` -- it turns out that serializing messages that use the standard Scala/Akka library types involve *far* more classes than one would expect -- for example, `Set` and `Map` involve a whole hive of specialized optimizations -- and that becomes a massive pain in the ass.  In fact, it's impossible to register some of them without cheating through *nasty* reflection, because a number of the classes you have to serialize are object-private.  Basically, `explicit` is pretty unusable with the Scala Standard Library in practice without it.

I created it out of sheer necessity, and it's designed for that explicit case, registering traits that would ordinarily never be involved in the process; I've given no thought to whether it is *safe* to use in the typical `automatic` case, because I don't use `automatic`.  It *might* work in the `automatic` case if used with care, but somebody's going to have to spend some serious time thinking that through and running some examples before I have any confidence in it.

In other words, I don't object in principle, but I kinda want to be convinced that it's not going to cause havoc...

**Comment by guidomedina** [comment]
I don't believe it is going to be dangerous as you do that for the listed classes only, not for the auto-registered, not having such functionality limited to a ID strategy instead of a phase limits it a lot and makes it too specific.

**Comment by guidomedina** [comment]
Automatic is very specific so it shouldn't make you less or more confident, it does exactly what it is described:

Register classes automatically that haven't been listed, and that happens AFTER the listed classes have been registered.

I'm not 100% sure but I think the incremental part doesn't send IDs but class names which makes it non-breaking, but the messages are bigger which is why people preferred explicit if possible.

**Comment by guidomedina** [comment]
I contributed something else to akka kryo but I'm a Java developer, I struggled with Scala :(

**Comment by jducoeur** [comment]
Hmm.  Well, like I said, I don't know `automatic` at all well.  Ultimately, this is @romix' call -- so long as somebody can demonstrate some decently thorough unit tests that show it working properly in the `automatic`-style environment (essentially, with `registrationRequired` turned off), I won't object to it.  But I'm not going to do the PR for it, since I don't feel like I understand it with enough confidence...

**Pull request merged**